kernel/ipi.cc

```diff
@@ -91,52 +91,76 @@ on_ipicall()
   }
 }
 
+/* Logic for implementing pause_other_cpus_and_call
+
+ The core that calls pause_other_cpus_and_call changes the
+ state from Ready to Pausing, Paused to Unpausing, and
+ Unpaused to Ready.
+
+ Of the remaining cores, the last one to pause/unpause will
+ change the state from Pausing to Paused and Unpausing to
+ Unpaused respectively.
+ */
 std::atomic<int> paused_cpu_counter {0};
-enum pause_state { Ready, Pausing, Paused, Unpausing, Unpaused };
-std::atomic<enum pause_state> pause_state {Ready};
-enum pause_state READY = Ready;
+enum pause_state_t { Ready, Pausing, Paused, Unpausing, Unpaused };
+std::atomic<pause_state_t> pause_state {Ready};
 
 // Use IPI to pause other cores until update is complete.
 void
 pause_other_cpus(void)
 {
-  while (!atomic_compare_exchange_strong(&pause_state, &READY, Pausing)); // wait for any previously paused CPUs to resume
+  // wait for any previously paused CPUs to resume
+  pause_state_t READY = Ready;
+  while (!pause_state.compare_exchange_weak(READY, Pausing)) {
+    // atomic::compare_exchange_weak modifies the expected variable
+    // in the failure case, so we need to change it back
+    READY = Ready;
+  }
```

Comment:
I'm I missing a reason you couldn't just pass _Ready_ directly to _compare_exchange_weak_?

Reply:
When I try using _Ready_ I get: `error: cannot bind non-const lvalue reference of type ‘pause_state_t&’ to an rvalue of type ‘pause_state_t’`. Based on the docs, _compare_exchange_weak_  expects the first argument to have a type of `T&`, which I think means that it should be a memory location that can be modified, whereas _Ready_ is a literal?

Reply:
Ah yes, didn't see that it needed a `T&` in the documentation